Allow adding custom highlight widget to custom pages in option

3nids opened the pull request:
try to fix #18032

I am getting a freeze in the line edit with this change.

@nirvn @nyalldawson can you try this with Qt 5.9 and see if it freezes when you type some text in the search line edit in global options?

I wonder if these freezes are not PyQt or SIP related. Because one (weird) place I have face one is the layer action dialog in vector layer properties (also when editing a line edit there).

3nids committed:
Allow adding custom highlight widget to custom pages in option

nyalldawson commented on src/gui/qgsoptionswidgetfactory.h:
```diff
@@ -59,6 +68,23 @@ class GUI_EXPORT QgsOptionsPageWidget : public QWidget
+    /**
+     * Register a highlight widget to be used to search and highlight text in
+     * options dialogs. This can be used to provide a custom implementation of
+     * QgsOptionsDialogHighlightWidget.
```
Can you add a note about ownership herem

nyalldawson commented on src/gui/qgsoptionsdialogbase.cpp:
```diff
@@ -256,7 +256,25 @@ void QgsOptionsDialogBase::registerTextSearchWidgets()
+      QgsOptionsPageWidget *opw = qobject_cast<QgsOptionsPageWidget *>( mOptStackedWidget->widget( i ) );
+      if ( opw )
+      {
+        customHighlightWidgets = opw->registeredHighlightWidgets();
```
I'm not clear about ownership here - are these widgets owned? Because the python bindings are transferring ownership yet I can't see where they are deleted.

nyalldawson commented:
I'll give it a test!

nyalldawson commented:
Freezes here to on 5.9

nyalldawson commented:
Removing the `self.adjustColumns()` from itemExpanded in ConfigDialog.py fixes this

nyalldawson committed:
Fix hang in options search widget

nyalldawson commented:
Cherry-pick this fix: https://github.com/nyalldawson/QGIS/commit/6d7d2d7caa12917da8f1a5f62f63de85c8d53c0e

One issue which needs addressing in the processing search is that it's not searching any items with children - so searching "buffer" returns nothing, because the matching hits have children. (Not caused by this PR)

3nids committed:
remove ownesrhip transfer

3nids committed:
do not search for less than 3 letters

3nids committed:
processing search: also search in group items and fix return value

3nids commented:
Thanks a lot for the patch and your comments.

I removed the Transfer. I thought it was required but apparently not.

I also fixed the search in group items.

The bad point is that now searching is very slow (for the 2-3 letters) as soon as the processing tree was shown once.

nyalldawson commented:
I've given this latest revision a test and it's working very nicely!

Two issues:
- I think when searching for processing items and a matching item with children is found, all its children should be shown too. It's what the "advanced" tab does and I think it's correct behaviour (since those children are settings related to the search term too)
- If I search for "digitizing" I think the WHOLE "digitizing" pane should be shown too!

3nids committed:
[processing] search also show children in config

3nids committed:
[options search] expand item when found in a tree
